### PR TITLE
Use a column to store categories, rather than a mapping

### PR DESCRIPTION
**Pull request by shwina**
This PR implements one of the changes mentioned in https://github.com/data-apis/dataframe-api/issues/41, i.e.,

> Add get_children() method, and store the mapping that is now in Column.describe_categorical in a child column instead. Note that child columns are also needed for variable-length strings.

Replacing the previous `mapping` with a child column now implies that the data buffer stores integer indices into the child column.

### CODE REVIEW

**Review by kkraus14 — APPROVED** [review]

**Comment by vnlitvinov** [comment]
@shwina @rgommers should we maybe not try to disguise the (explicit) categories as generic "children"?

I think just replacing the _type_ of `df.describe_categorical['mapping']` from `dict` to `Column` instead of introducing of a new method could be better - it's still a mapping after all 🙃

**Comment by rgommers** [comment]
We briefly touched on this PR today, we'd like to move it forward since all interchange protocol implementers seem to be in favor of this. @vnlitvinov plans to revisit his proposal from above and add that here or open an alternative PR. Then we can try to finalize this, and then update implementations in the various libraries for it.

**Comment by shwina** [comment]
> still a mapping after all

I'm not sure I quite follow; the `Column` is explicitly _not_ a mapping, rather it is concretely the `categories`. Perhaps I'm missing something?

**Comment by vnlitvinov** [comment]
@shwina I've missed your comment, so I went ahead with the original plan we discussed last meeting and I had added a commit doing so. If you don't like the idea feel free to remove my commit.

As for it not being a mapping... it somewhat is :) it maps an integer index to a category value.

**Comment by rgommers** [comment]
After discussion on Thursday, @shwina and others were happy with the changes @vnlitvinov pushed. With one change to make: rename `mapping` to `categories`, since it's no longer a real mapping (only implicitly, the keys are `0 ... n_categories`) and it would help with transitioning too.

File: protocol/dataframe_protocol.py [hunk 1]
```diff
@@ -170,23 +170,24 @@ def dtype(self) -> Tuple[enum.IntEnum, int, str, str]:
...
 
     @property
-    def describe_categorical(self) -> dict[bool, bool, Optional[dict]]:
+    def describe_categorical(self) -> dict[bool, bool, Optional[Column]]:
         """
         If the dtype is categorical, there are two options:
 
         - There are only values in the data buffer.
-        - There is a separate dictionary-style encoding for categorical values.
+        - There is a separate non-categortical Column encoding categorical values.
```
Comment (kkraus14) [inline protocol/dataframe_protocol.py:178]:
```suggestion
        - There is a separate non-categorical Column encoding categorical values.
```

Comment (vnlitvinov) [reply protocol/dataframe_protocol.py]: nice catch, thanks!

**Comment by rgommers** [comment]
> With one change to make: rename `mapping` to `categories`,

This change was made in the spec part, but not the rest of this PR. I'm inclined to push a change to `pandas_implementation.py` and then merge this.

**Comment by rgommers** [comment]
> > With one change to make: rename `mapping` to `categories`,
> 
> This change was made in the spec part, but not the rest of this PR. I'm inclined to push a change to `pandas_implementation.py` and then merge this.

I tried this - since the tests no longer pass with this PR, and we now have an _actual_ Pandas implementation, it may be better to completely remove this early prototype in `pandas_implementation.py`.

**Comment by rgommers** [comment]
> I tried this - since the tests no longer pass with this PR, and we now have an _actual_ Pandas implementation, it may be better to completely remove this early prototype in `pandas_implementation.py`.

Discussed in today's call: everyone is happy with deleting this prototype code.

**Review by rgommers — APPROVED** [review]
Everyone is happy and this is now a pretty small/straightforward diff. So merging. I'll follow up with all the implementers to ensure we will actually propagate these changes to all implementations.

Thanks @shwina, @vnlitvinov, @kkraus14

**Pull request merged**